testing/opae-c/test_feature_enum_c.cpp

```diff
@@ -0,0 +1,257 @@
+	std::array<fpga_token, 2> tokens_;
+	uint32_t num_matches_;
+	test_platform platform_;
+	test_device invalid_device_;
+	test_system *system_;
+};
+
+/**
+ * @test       test_feature_enumerate
+ * @brief      Tests: fpgaFeatureEnumerate, xfpga_fpgaCloneFeatureToken
+ * @details    When fpgaFeatureEnumerate() is called with a valid param,<br>
+ *             then it enumerates the accelerateor's DFH to find specific<br>
+ *             type of BBB specified in the feature filter<br>
+ *             When xfpga_fpgaCloneFeatureToken() is called with a valid param,<br>
+ *             fpga_feature_token is cloned.<br>
+ */
+TEST_P(opae_feature_enum_c_p, test_feature_enumerate) {
+
+	struct DFH dfh ;
+	dfh.id = 0x1;
+	dfh.revision = 0;
+	dfh.next_header_offset = 0x100;
+	dfh.eol = 1;
+	dfh.reserved = 0;
+	dfh.type = 0x1;
+
+	EXPECT_EQ(FPGA_OK, fpgaWriteMMIO64(accel_, which_mmio_, 0x0, dfh.csr));
+
+	EXPECT_EQ(FPGA_OK, fpgaWriteMMIO64(accel_, which_mmio_, 0x8, 0xf89e433683f9040b));
+	EXPECT_EQ(FPGA_OK,fpgaWriteMMIO64(accel_, which_mmio_, 0x10, 0xd8424dc4a4a3c413));
+
+	struct DFH dfh_bbb = { 0 };
+
+	dfh_bbb.type = 0x2;
+	dfh_bbb.id = 0x2;
+	dfh_bbb.revision = 0;
+	dfh_bbb.next_header_offset = 0x000;
+	dfh_bbb.eol = 1;
+	dfh_bbb.reserved = 0;
+
+	EXPECT_EQ(FPGA_OK, fpgaWriteMMIO64(accel_, which_mmio_, 0x100, dfh_bbb.csr));
+
+	EXPECT_EQ(FPGA_OK, fpgaWriteMMIO64(accel_, which_mmio_, 0x108, 0x9D73E8F258E9E3E7));
+	EXPECT_EQ(FPGA_OK, fpgaWriteMMIO64(accel_, which_mmio_, 0x110, 0x87816958C1484CE0));
+
+	EXPECT_EQ(fpgaFeatureEnumerate(accel_, &feature_filter_, NULL,
+		0, &num_matches_), FPGA_OK);
+
+	EXPECT_EQ(fpgaFeatureEnumerate(accel_, &feature_filter_, ftokens_.data(),
+		ftokens_.size(), &num_matches_), FPGA_OK);
+
+	EXPECT_EQ(fpgaDestroyFeatureToken(&(ftokens_[0])), FPGA_OK);
+}
+
+/**
+ * @test       nullfilter
+ * @brief      Tests: fpgaFeatureEnumerate
+ * @details    When fpgaFeatureEnumerate() is called with an invalid<br>
+ *             feature property, then it will return FPGA_INVALID_PARAM.<br>
+ */
+TEST_P(opae_feature_enum_c_p, nullfilter) {
+	EXPECT_EQ(fpgaFeatureEnumerate(accel_, nullptr, ftokens_.data(), ftokens_.size(), &num_matches_),
+            FPGA_INVALID_PARAM);
+}
+
+/**
+ * @test       nullmatches
+ * @brief      Tests: fpgaFeatureEnumerate
+ * @details    When fpgaFeatureEnumerate() is called with null for num_matches,<br>
+ *             it will return FPGA_INVALID_PARAM.<br>
+ */
+TEST_P(opae_feature_enum_c_p, nullmatches) {
+	EXPECT_EQ(fpgaFeatureEnumerate(accel_, &feature_filter_, ftokens_.data(), ftokens_.size(), NULL),
+            FPGA_INVALID_PARAM);
+}
+
+/**
+ * @test       nulltokens
+ * @brief      Tests: fpgaFeatureEnumerate
+ * @details    When fpgaFeatureEnumerate() is called with null for tokens,<br>
+ *             it will return FPGA_INVALID_PARAM.<br>
+ */
+TEST_P(opae_feature_enum_c_p, nulltokens) {
+	EXPECT_EQ(fpgaFeatureEnumerate(accel_, &feature_filter_, NULL, ftokens_.size(), &num_matches_),
```

Comment:
I think this result also depends on ftokens_.size() being non-zero.  Need to mention that in the comments above.

Reply:
Done

Reply:
NULL tokens and non-zero ftokens_.size(), I think?  Can you document the requirements?

Reply:
Sure.